packages/core-p2p/src/socket-server/controllers/peer.ts

```diff
@@ -0,0 +1,202 @@
+        ) {
+            throw new TooManyTransactionsError(deserializedHeader.data);
+        }
+
+        const deserialized: {
+            data: Interfaces.IBlockData;
+            transactions: Interfaces.ITransaction[];
+        } = Blocks.Deserializer.deserialize(blockHex);
+
+        const block: Interfaces.IBlockData = {
+            ...deserialized.data,
+            transactions: deserialized.transactions.map((tx) => tx.data),
+        };
+
+        const fromForger: boolean = isWhitelisted(
+            configuration.getOptional<string[]>("remoteAccess", []),
+            request.info.remoteAddress,
+        );
+
+        const blockchain = this.app.get<Contracts.Blockchain.Blockchain>(Container.Identifiers.BlockchainService);
+
+        if (!fromForger) {
+            if (blockchain.pingBlock(block)) {
+                return true;
+            }
+
+            const lastDownloadedBlock: Interfaces.IBlockData = blockchain.getLastDownloadedBlock();
+
+            if (!Utils.isBlockChained(lastDownloadedBlock, block)) {
+                throw new UnchainedBlockError(lastDownloadedBlock.height, block.height);
+            }
+        }
+
+        if (
+            block.transactions &&
+            block.transactions.length > Managers.configManager.getMilestone().block.maxTransactions
+        ) {
+            throw new TooManyTransactionsError(block);
+        }
+
+        this.logger.info(
+            `Received new block at height ${block.height.toLocaleString()} with ${Utils.pluralize(
+                "transaction",
+                block.numberOfTransactions,
+                true,
+            )} from ${request.info.remoteAddress} (${request.headers.host})`,
+        );
+
+        blockchain.handleIncomingBlock(block, fromForger);
+        return true;
+    }
+
+    public async postTransactions(request: Hapi.Request, h: Hapi.ResponseToolkit): Promise<string[]> {
+        this.validatePayload(request.payload, peerSchemas.postTransactions);
+        
+        // Add the peer if not already (on every peer endpoint)
+        this.peerProcessor.validateAndAcceptPeer({ ip: request.info.remoteAddress } as Contracts.P2P.Peer);
+
+        const createProcessor: Contracts.TransactionPool.ProcessorFactory = this.app.get(
+            Container.Identifiers.TransactionPoolProcessorFactory,
+        );
+        const processor: Contracts.TransactionPool.Processor = createProcessor();
+        await processor.process((request.payload as any).transactions);
+        return processor.accept;
+    }
+
+    public async getBlocks(
+        request: Hapi.Request,
+        h: Hapi.ResponseToolkit,
+    ): Promise<Interfaces.IBlockData[] | Contracts.Shared.DownloadBlock[]> {
+        this.validatePayload(request.payload, peerSchemas.getBlocks);
+        
+        // Add the peer if not already (on every peer endpoint)
+        this.peerProcessor.validateAndAcceptPeer({ ip: request.info.remoteAddress } as Contracts.P2P.Peer);
```

Comment:
Could this be moved into a plugin that is responsible for peer acceptance? Seems to be duplicated in every controller method.

Reply:
done.